### DISCUSSION TITLE
Use SageMath with Jupyter

### DISCUSSION THREAD

**Discussion by TS-CUBED** [quarto-dev/quarto-cli#4231] [Q&A] [answered]
SageMath (https://sagemath.org)  is a Mathematics Package, based on Python (and others), which comes with it's own Jupyter kernel. Would it be possible to add support for that in Quarto?

**Comment by cscheid** [comment, marked as answer]
> which comes with it's own Jupyter kernel

If that's the case, then you should be able to say

```yaml
jupyter:
  kernel: sage-math
```

(or whatever the name of the kernel is) and then use it.

**Reply by TS-CUBED** [reply]
So do I understand this as: anything that is supported by Jupyter will work in Quarto?

**Reply by TS-CUBED** [reply]
Indeed. I had the wrong name (didn't check the kernelspecs).

In the YAML header:
```
jupyter:
  kernel: sagemath
``` 

And then the codeblocks:

    ```{sage}
    
    ```

I had `sagemath` as the language name (which is what org-babel uses), but it needs to be `sage` (kernel and language name are different).

Brilliant!

**Reply by cscheid** [reply]
> So do I understand this as: anything that is supported by Jupyter will work in Quarto?

That's the intention, yes. We can't test every single kernel out there, but in general we treat kernels that work in jupyter but not in quarto as quarto bugs.

**Comment by erentar** [comment]
How can i tell what is the kernel name, language name pair for my language?

**Reply by mcanouil** [reply]
Simply list the kernels using Jupyter.

```bash
jupyter kernelspec list
```

**Reply by TS-CUBED** [reply]
That will list the kernel name but that's not the same as the label used in the code block. E.g. for Julia my kernel name is `julia-1.10` but it will require `{julia}` in the codeblock name; the kernel `sagemath` has the label `{sage}` so it's not straightforward to find which label is required for which language.

**Reply by mcanouil** [reply]
The kernel documentation does not document that?

I am pretty sure, the language name is defined within the kernel.

**Reply by TS-CUBED** [reply]
It is indeed. But `kernelspec list` won't tell.

@erentar: To find that use

```
jupyter kernelspec list
```

to find the name of the kernel, e.g.:

```
$ jupyter kernelspec list
Available kernels:
  %s    %s julia-1.10 /home/thor/.local/share/jupyter/kernels/julia-1.10
  %s    %s python3    /usr/share/jupyter/kernels/python3
  %s    %s sagemath   /usr/share/jupyter/kernels/sagemath
```

then look in the kernel definition `kernel.json` file for the defined language name:

```
$ cat /usr/share/jupyter/kernels/sagemath/kernel.json 
{"argv": ["/usr/bin/sage", "--python", "-m", "sage.repl.ipython_kernel", "-f", "{connection_file}"], "display_name": "SageMath 9.5", "language": "sage"}
```

So in this case `kernelspec list` tells me that the kernel name is `sagemath`, and `kernel.json` tells me that the language name is `sage`.
